docker/docker_client.go

```diff
@@ -244,37 +248,96 @@ func newDockerClient(sys *types.SystemContext, registry, reference string) (*doc
 	if err := tlsclientconfig.SetupCertificates(certDir, tr.TLSClientConfig); err != nil {
 		return nil, err
 	}
+	tr.TLSClientConfig.InsecureSkipVerify = sysContextSkipVerify(sys)
+
+	// Find the registry
+	reg, err := sysregistriesv2.FindRegistry(sys, reference)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error loading registries")
+	}
+	if reg != nil {
+		// Retrieve the docker client from the found registry
+		return dockerClientFromSysRegistry(ctx, sys, reg, tr), nil
+	}
 
+	// No registry found within the configuration, fallback to the provided
+	// one
+	return &dockerClient{
+		registry:              registry,
+		sys:                   sys,
+		client:                &http.Client{Transport: tr},
+		insecureSkipTLSVerify: tr.TLSClientConfig.InsecureSkipVerify,
+	}, nil
+}
+
+func sysContextSkipVerify(sys *types.SystemContext) bool {
 	// Check if TLS verification shall be skipped (default=false) which can
 	// either be specified in the sysregistriesv2 configuration or via the
 	// SystemContext, whereas the SystemContext is prioritized.
-	skipVerify := false
 	if sys != nil && sys.DockerInsecureSkipTLSVerify != types.OptionalBoolUndefined {
 		// Only use the SystemContext if the actual value is defined.
-		skipVerify = sys.DockerInsecureSkipTLSVerify == types.OptionalBoolTrue
-	} else {
-		reg, err := sysregistriesv2.FindRegistry(sys, reference)
-		if err != nil {
-			return nil, errors.Wrapf(err, "error loading registries")
+		return sys.DockerInsecureSkipTLSVerify == types.OptionalBoolTrue
+	}
+	return false
+}
+
+// dockerClientFromSysRegistry returns a `*dockerClient` for the provided
+// sysregistriesv2. If the registry contains any mirrors, it will be tried
+// to detect their properties. The mirror will be used if the property
+// detection succeeds. If not, the fallback registry URL will be used instead.
+func dockerClientFromSysRegistry(
+	ctx context.Context,
+	sys *types.SystemContext,
+	registry *sysregistriesv2.Registry,
+	transport *http.Transport,
+) *dockerClient {
+	// registry != nil should already be checked by referring functions
+	// Check for available mirrors
+	for _, mirror := range registry.Mirrors {
+		// Adapt the InsecureSkipVerify only when not already set from the
+		// system context
+		if !transport.TLSClientConfig.InsecureSkipVerify {
```

Comment:
This look suspect; does it handle `DockerInsecureSkipTLSVerify: OptionalBoolFalse` correctly? (I didn’t check in detail)

Reply:
It should, first of all we create the transport in `newDockerClient()`:

https://github.com/containers/image/blob/e8c87562e0c7b850e26947a81eb03382495b60ca/docker/docker_client.go#L236

then we set the initial `InsecureSkipVerify`:

https://github.com/containers/image/blob/e8c87562e0c7b850e26947a81eb03382495b60ca/docker/docker_client.go#L251
https://github.com/containers/image/blob/e8c87562e0c7b850e26947a81eb03382495b60ca/docker/docker_client.go#L273-L282

After that the `transport.TLSClientConfig.InsecureSkipVerify` can only be modified when `false`, since the system context gets more priority. I also thought about using a bool pointer here but it would not make it easier to understand. All three functions rely tightly on each other which makes it also fragile to modifications. I will check if I can do something here...

Reply:
The point is that DockerInsecureSkipTLSVerify` set to _both_ `OptionalBoolFalse` and `OptionalBoolUndefined` causes `sysContextSkipVerify` to return `false`, and then `if !transport.TLSClientConfig.InsecureSkipVerify` can’t distinguish in between the two.

Reply:
… and, to add context, the idea is that `SystemContext.DockerInsecureSkipTLSVerify` takes precedence over the `sysregistriesv2` config file.

Reply:
Okay I think I got it and will adapt it as needed.

Reply:
This point should be solved with the new approach.